shared/ui/textField/TextField.tsx

```diff
@@ -1,102 +1,123 @@
+// To be able to pass these props, you can extract the desired union member using the Extract utility type.
+// https://www.typescriptlang.org/docs/handbook/utility-types.html#extracttype-union
 export type TextFieldProps =
-  // <T extends TextFieldType = 'input'>
+  | (SharedProps & {
+      as?: 'input';
+      inputType?: 'password' | 'search' | 'text';
+    })
+  | (SharedProps & { as: 'textarea' });
 
-  {
-    as?: 'input' | 'textarea';
-    className?: string;
-    disabled?: boolean;
-    errors?: string;
-    inputtype?: 'password' | 'text';
-    isSearchInput?: boolean;
-    label?: string;
-    onChange: (e: string) => void;
-    onFocus?: () => void;
-    placeholder?: string;
-    required?: boolean;
-    value: string;
-  };
+// Forces the 'ref' type match the 'as' prop type. You should not specify the 'as' props dynamically.
+// Use only the literal value to avoid possible errors with non-existent properties of the target element pointed by the 'ref'.
+// prettier-ignore
+export type TextFieldComponent = {
+  (props: Extract<TextFieldProps, { as: 'textarea'}> & { ref?: ForwardedRef<HTMLTextAreaElement> }): ReactNode;
+  (props: Extract<TextFieldProps, { as?: 'input'}> & { ref?: ForwardedRef<HTMLInputElement> }): ReactNode;
+}
 
-export const TextField = forwardRef((props: TextFieldProps, ref) => {
-  const {
-    as,
-    className,
-    disabled,
-    errors,
-    inputtype = 'text',
-    isSearchInput,
-    label,
-    onChange,
-    placeholder = 'email',
-    required,
-    value,
-  } = props;
+export const TextField: TextFieldComponent = forwardRef<
+  HTMLInputElement | HTMLTextAreaElement,
+  TextFieldProps
+>(({ className, error, label, onChange, required, ...props }, ref) => {
+  const [showPassword, setShowPassword] = useState(false);
 
-  const [type, setType] = useState(inputtype);
+  const handleChange = (e: ChangeEvent<HTMLInputElement | HTMLTextAreaElement>) =>
+    onChange(e.currentTarget.value);
 
-  const changeInputType = () => {
-    type === 'password' ? setType('text') : setType('password');
-  };
+  const sharedClassName = clsx(s.input, error && s.error, className);
+  let resolvedFragment: ReactNode;
 
-  const clearTextField = () => {
-    onChange('');
-  };
+  // Do not destructure the 'as' prop before 'if' statement
+  // because TS will lose the connection between the 'as' prop and type of the remaining props
+  if (!props.as || props.as === 'input') {
+    const { as, inputType, ...restProps } = props;
+    const isSearch = inputType === 'search';
+    const isPassword = inputType === 'password';
+    const passwordFieldNextType = showPassword ? 'text' : 'password';
 
-  const { as: Component = 'input', ...rest } = props;
+    resolvedFragment = (
+      <>
+        <input
+          className={clsx(sharedClassName, isSearch && s.inputSearch)}
+          onChange={handleChange}
+          ref={ref as unknown as ForwardedRef<HTMLInputElement>}
```

Comment:
Hack to satisfy Typescript. Since we make sure when rendering a component using the `TextFieldComponent` type that the `ref` type is passed correctly, we can cast the type here via the `as` keyword, instead of explicitly checking for the correct `ref` type by cluttering up the code.